File: indexold.js
function simulateKeyFindingRuns(runs) {
    const BOARD_SIZE = 14;
    let totalMoves = 0;

    // Helper function to get a random integer in the range [0, max)
    function getRandomInt(max) {
        return Math.floor(Math.random() * max);
    }

    // Function to generate a new board
    function generateBoard() {
        let board = Array(BOARD_SIZE).fill(null);

        // Decide if there's a key (50% chance)
        let hasKey = Math.random() < 0.5;
        if (hasKey) {
            board[getRandomInt(BOARD_SIZE)] = 'key';
        }

        // Place a shuffle in a random box
        let shuffleIndex;
        do {
            shuffleIndex = getRandomInt(BOARD_SIZE);
        } while (board[shuffleIndex] === 'key');
        board[shuffleIndex] = 'shuffle';

        // Place the uncover in a random box
        let uncoverIndex;
        do {
            uncoverIndex = getRandomInt(BOARD_SIZE);
        } while (board[uncoverIndex] === 'key' || board[uncoverIndex] === 'shuffle');
        board[uncoverIndex] = 'uncover';

        return board;
    }

    // Function to play a run until a key is found
    function playRun(board) {
        let moves = 0;

        while (true) {
            // Open a random box
            let boxIndex = getRandomInt(BOARD_SIZE);

            if (board[boxIndex] === 'key') {
                // Key found, end the run
                moves++;
                return moves;
            } else if (board[boxIndex] === 'shuffle') {
                // Shuffle found, reset board and continue
                board = generateBoard();
                moves++;
            } else if (board[boxIndex] === 'uncover') {
                // Uncover two other boxes
                let uncovered = 0;
                while (uncovered < 2) {
                    let randomIndex = getRandomInt(BOARD_SIZE);
                    if (randomIndex !== boxIndex && board[randomIndex] !== null) {
                        uncovered++;
                        if (board[randomIndex] === 'key') {
                            // Key found through uncover, end the run
                            moves++;
                            return moves;
                        } else if (board[randomIndex] === 'shuffle') {
                            // Shuffle found through uncover, reset board and continue
                            board = generateBoard();
                            moves++;
                            break;
                        }
                    }
                }
                moves++;
            } else {
                // Empty box opened
                moves++;
            }
        }
    }

    // Simulate multiple runs to find the average number of moves
    for (let i = 0; i < runs; i++) {
        let board = generateBoard();
        totalMoves += playRun(board);
    }

    return totalMoves / runs;
}

// Simulate 1000 runs and print the average number of moves to find a key
console.log(simulateKeyFindingRuns(1000000));

File: indexold.py
import random

def simulate_board_with_currency(trials):
    total_boxes_opened = 0
    total_bells_spent = 0

    for _ in range(trials):
        boxes_opened = 0
        bells_spent = 0

        while True:
            # Step 1: 50% chance there is a key
            has_key = random.random() < 0.5
            key_position = random.randint(0, 13) if has_key else None

            # Step 2: Randomly place the shuffle
            shuffle_position = random.randint(0, 13)

            # Step 3: Randomly assign a bell reward (2/3 chance)
            bell_reward_position = random.randint(0, 13) if random.random() < (2/3) else None

            # Step 4: Randomly choose a box that will reveal two other boxes
            reveal_box_index = random.randint(0, 13)
            revealed_boxes = []

            while len(revealed_boxes) < 2:
                idx = random.randint(0, 13)
                if idx != reveal_box_index and idx not in revealed_boxes:
                    revealed_boxes.append(idx)

            # Handle box opening logic - IDENTICAL to your original logic
            boxes_opened += 1  # The reveal box counts as opened

            if key_position in revealed_boxes:
                # We found the key in the revealed boxes
                boxes_opened += 1  # Open the box with the key
                break
            elif shuffle_position in revealed_boxes:
                # Shuffle hit, restart with a new board
                continue  # Go to the next board
            else:
                # None of the revealed boxes have the key or shuffle
                boxes_opened += 2  # Open the revealed boxes and continue searching

                # Open the rest of the boxes if we haven't found the key yet
                for i in range(14):
                    if i not in revealed_boxes and i != reveal_box_index:
                        boxes_opened += 1
                        if i == key_position:
                            break  # Stop when we find the key
            break  # End this trial if we don't hit shuffle

        total_boxes_opened += boxes_opened

        # Now, calculate the bells spent based on the number of boxes opened
        # (Currency logic starts after we determine how many boxes were opened)

        # Step 1: Open the reveal box (net -7 bells)
        bells_spent += 10  # Opening the reveal box costs 10 bells
        bells_spent -= 3   # You get 3 bells back (net -7 bells)

        # Step 2: Process the two revealed boxes if opened
        bells_spent += 10 * 2  # 10 bells for each of the two revealed boxes

        # Step 3: Check if any revealed boxes contain a bell reward (gain 14 bells)
        if bell_reward_position in revealed_boxes:
            bells_spent -= 14  # Gain 14 bells from the bell reward box

        # Step 4: Process remaining boxes
        for i in range(14):
            if i not in revealed_boxes and i != reveal_box_index:
                bells_spent += 10  # 10 bells to open each box

                # If the current box is the bell reward box, adjust the bell count
                if i == bell_reward_position:
                    bells_spent -= 14  # Gain 14 bells from the bell reward

                if i == key_position:
                    break  # Stop processing once the key is found

        total_bells_spent += bells_spent

    # Return the average number of boxes opened and bells spent per key
    average_boxes_opened = total_boxes_opened / trials
    average_bells_spent = total_bells_spent / trials
    return average_boxes_opened, average_bells_spent

# Simulate over a large number of trials
trials = 1000000
average_boxes_opened, average_bells_spent = simulate_board_with_currency(trials)
print(f"Average number of boxes opened to find a key: {average_boxes_opened:.2f}")
